### DISCUSSION TITLE
Allow reading from environment variables in configuration file

### DISCUSSION THREAD

**Discussion by poperigby** [Kozea/Radicale#1565] [Ideas]
This would be useful for reading secrets in the configuration file. Environment variables could be written like this:

```
key: {{VARIABLE}}
```

**Comment by pbiering** [comment]
> This would be useful for reading secrets in the configuration file. Environment variables could be written like this:
> 
> ```
> key: {{VARIABLE}}
> ```

this is not a format of a Windows or Linux environment option...it must be something like that

```
RADICALE_OPTION__<section>__<key>="<value>"
```

and order should be config file -> environment -> command line

Contribution is required

**Reply by poperigby** [reply]
> this is not a format of a Windows or Linux environment option...it must be something like that

I was talking about having the environment variable directly in the config file, but that would work too.

I could take a crack at it. Do you mind pointing me in the right direction?

**Reply by pbiering** [reply]
> > this is not a format of a Windows or Linux environment option...it must be something like that
> 
> I was talking about having the environment variable directly in the config file, but that would work too.

Hmm, also an option.

> I could take a crack at it. Do you mind pointing me in the right direction?

`radicale/config.py` would require an extension.

**Reply by poperigby** [reply]
Thanks. What would be your preference? In the config file, or not?

**Reply by pbiering** [reply]
> Thanks. What would be your preference? In the config file, or not?

I would prefer having support without tweaking the config file, so imply lookup environment, search for entries matching the pattern

`RADICALE_OPTION__<section>__<key>`

validate `<value>` and then override the internal one provided by config file and/or command line option, I would also prefer having this logged on info level.

**Reply by poperigby** [reply]
Sounds good. Where should I put this environment variable handling in `config.py`?

**Reply by pbiering** [reply]
I would assume the `update` function must be somehow extended or cloned+adjusted...as it is called currently by `__main__.py`

```
   try:
        configuration = config.load(config.parse_compound_paths(
            config.DEFAULT_CONFIG_PATH,
            os.environ.get("RADICALE_CONFIG"),
            os.pathsep.join(args_ns.config) if args_ns.config is not None
            else None))
        if arguments_config:
            configuration.update(arguments_config, "command line arguments")
    except Exception as e:
        logger.critical("Invalid configuration: %s", e, exc_info=True)
        sys.exit(1)
```

Before `if arguments_config` the environment based configuration should be called imho.

**Comment by pbiering** [comment]
Draft: https://github.com/Kozea/Radicale/pull/1577
